[tftrt update]

jjsjann123 opened the pull request:
Added support for TRT plugin during conversion
  - converter & shape inference are now aware of plugin factory.
  - PluginTensorRT
    - wrapper for nvinfer1::IPlugin
    - base class for custom plugins
    - provides serialization/deserialization of plugin op type & input tensor dimensions
  - PluginFactoryTensorRT
    - wrapper for nvinfer1::PluginFactory
    - singleton with plugin registration, allow user to register plugin op type (string), plugin construction function & plugin deserialization function.
    - owns constructed & deserialized plugins.
  - Unit tests for plugin registration

  * compatible with TRT 3.0.4 plugin API.
  * future plugin API changes willl be updated.

jjsjann123 committed:
[tftrt update]
  Added support for TRT plugin during conversion
  - converter & shape inference are now aware of plugin factory.
  - each plugin does serialization of plugin type & input dimensions
  - wrapper for nvinfer1::IPlugin & nvinfer1::PluginFactory

  * compatible with TRT 3.0.4 plugin API.
  * future plugin API changes willl be updated.

jjsjann123 commented:
@aaroey

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.h:
```diff
@@ -0,0 +1,83 @@
+  void DestroyPlugins();
+
+ protected:
+  std::unordered_map<size_t,
```
Can we use the plugin name string as the key? Using a hash code makes things much harder to understand, so any reason doing that?

jjsjann123 replied:
As said earlier, serialization & registration part is merely a placeholder here waiting for updated TRT API.
But I could switch to string other than hash code for this one here
 :)

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.cc:
```diff
@@ -0,0 +1,81 @@
+  instance_m_.lock();
+  auto plugin_ptr =
+      plugin_registry_[encoded_op_name].first(serial_data, serial_length);
+  // string op_name = "IncPluginTRT";
```
Remove these?

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.cc:
```diff
@@ -0,0 +1,81 @@
+  return ret.second;
+}
+
+void PluginFactoryTensorRT::DestroyPlugins() { return; }
```
Why not destroying?

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.cc:
```diff
@@ -0,0 +1,81 @@
+namespace tensorflow {
+namespace tensorrt {
+
+PluginTensorRT* PluginFactoryTensorRT::createPlugin(const char* layerName,
```
Please fix all the variable names.

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.h:
```diff
@@ -0,0 +1,83 @@
+  // construction
+  PluginTensorRT* CreatePlugin(const string* op_name);
+
+  static PluginFactoryTensorRT& GetInstance() {
```
I would prefer to return a pointer here, so user won't accidentally make a copy.

aaroey commented on tensorflow/contrib/tensorrt/BUILD:
```diff
@@ -272,3 +275,26 @@ tf_cc_test(
 )
+
+# Library for the plugin factory 
+#cc_library(
```
Remove this line?

aaroey commented on tensorflow/contrib/tensorrt/BUILD:
```diff
@@ -67,6 +67,7 @@ tf_cuda_library(
     visibility = ["//visibility:public"],
     deps = [
         ":trt_logging",
+        ":trt_plugins",
```
I think there will be double import problems here: the plugin library has a singleton, but will be imported twice by two different python library (trt_convert_py and trt_ops_py), so at the end there will be two different singletons. @samikama to confirm.

aaroey replied:
After some analysis, I'm pretty sure this will cause problem:

1. the plugin factory is a global instance
2. it is linked here by trt_shape_function, which will eventually be linked into python/ops/_trt_engine_op.so which will be loaded by the python library
3. it is also linked into trt_engine_op_kernel and trt_conversion, which will eventually linked into _wrap_conversion.so which will be loaded by the python library
4. so there will be two copies of the factory, which is dangerous and will cause problems

I would suggest to make a header-only library and let trt_shape_function depend on it instead. The Logger class in the trt_logging rule also has this problem, but since it doesn't contain any data, it should be fine (but I still want to fix it later).

jjsjann123 replied:
I thought linux dynamic linking would be able to resolve duplicate libraries at runtime, not sure how it would perform on other os neither.
But I'm not very good with these stuff and the solution sounds straightforward. Will update and ping you.

jjsjann123 replied:
So currently trt_shape_function requires plugin factory to deserialize the engine; so does trt_conversion.
It does not seem clear to me how I can turn plugin factory into a header-only library to break the potential double import. It would be great if you could provide more details :)

Technically speaking, I can change trt_shape_function to use attributes to store output tensor shapes (required through TF static shape inference during segmentation) and avoid engine deserialization. 
This would work better with our optimization pass & last minute conversion during execution.

jjsjann123 replied:
Quick update:
We should remove the deserialization in trt_shape_function. Sami is also doing the same thing in his current PR where he's modifying the trt_engine_op. To avoid conflict I'll wait for his PR to be merged and then merge on top of that.

aaroey replied:
My understanding is that linker will load the first copy of the factory instance and ignore the second, regardless which .so is loaded first. Then, if the one who is not linked against the plugin registration is loaded first (maybe _trt_engine_op.so), then later when it tried to find the plugin it will fail. I could be wrong though.

We can see that both .so file has GetInstance() defined as a unique global symbol.

aaroey replied:
Thanks J, just saw your updates, that works for me. Meanwhile I'm fixing the formatting issues.

jjsjann123 replied:
Was digging a bit more into this. Since we will remove the code, so it's not related anymore.

I start to think this might not be an issue. Member function from an ordinary class has external linkage and could be accessed in other translation units. http://en.cppreference.com/w/cpp/language/storage_duration 
I tested it locally on with gcc. 
I have two shared libraries (lib1.so, lib2.so) all linked against a singleton library (libstg.so) . My client application could access the singleton through interfaces exposed in lib1.so & lib2.so. I think dynamic linker takes the first encounter of the getInstance and uses it for both (Windows might behave differently tho).

One side thing I caught is that I didn't guard my singleton initialization between threads. Will update that.

aaroey commented on tensorflow/contrib/tensorrt/BUILD:
```diff
@@ -272,3 +275,26 @@ tf_cc_test(
+    ],
+    linkstatic = 1,
+    deps = [
+        #"@protobuf_archive//:protobuf_headers",
```
Remove? Or you may really want to use protobuf as suggested below.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -246,6 +247,15 @@ class TFAttrs {
 
+  std::vector<string> GetAllAttrKey() {
+    std::vector<string> attr_list;
+    for (AttrMap::iterator iter = attrs_.begin(); iter != attrs_.end();
```
How about using range-based for loop?

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1173,6 +1194,43 @@ tensorflow::Status BinaryTensorOpTensor(
+  TFAttrs attrs(node_def);
+  auto attr_key_vector = attrs.GetAllAttrKey();
+  for (auto attr_key : attr_key_vector) {
+    std::cout << attr_key << std::endl;
```
VLOG?

jjsjann123 commented on tensorflow/contrib/tensorrt/plugin/trt_plugin.cc:
```diff
@@ -0,0 +1,89 @@
+
+PluginTensorRT::PluginTensorRT(const void* serialized_data, size_t length) {
+  // sanity check.
+  assert(EncodeOpName(GetPluginName()) !=
```
The assert does not really matter here, since it has already been checked by PluginFactoryTensorRT::createPlulgin(...)

I should have used ExtractOpName, since I'm trying to consume the first couple bytes of serializaed_data inserted by the serialize().
I'm doing this simply because of current TRT plugin API. I could wrap it differently to pass serialized_data as (void**) to allow data consumption marked by each deserialize function call.

PluginTensorRT is an abstract class. We could have multiple child-classes utilizing the same serialize() & parent constructor. Each would have different op name that was throw at the serialized_data.

jjsjann123 committed:
[PR comment addressed]
  adding plugin test for registration
  updating plugin API wrapper addressing comments in the PR
  addressing coding style issues
  removing commented code

jjsjann123 commented:
- Updated the description.
- Added unit test.
Let me know if we want more details & examples here @aaroey .

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.cc:
```diff
@@ -37,42 +37,46 @@ PluginTensorRT* PluginFactoryTensorRT::createPlugin(const char* layerName,
 
-void PluginFactoryTensorRT::DestroyPlugins() { return; }
+void PluginFactoryTensorRT::DestroyPlugins() {
+  instance_m_.lock();
```
Use std::lock_guard here and above?

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_plugin.cc:
```diff
@@ -0,0 +1,89 @@
+  count *= batchSize;
+  const float *input = reinterpret_cast<const float*>(inputs[0]);
+  float *output = reinterpret_cast<float*>(outputs[0]);
+  IncrementKernel(input, inc_, output, count, stream);
```
Where is this defined?

jjsjann123 commented on tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_plugin.cc:
```diff
@@ -0,0 +1,89 @@
+
+#if GOOGLE_CUDA
+#if GOOGLE_TENSORRT
+#include "inc_op_plugin.h"
```
Oops. I think I pushed the wrong commit here. The example is used for my plugin test (making sure things worked).
I'll clean up the code and update it.

clang-format hasn't been run on this one, neither has I done any style guide check. Will update the code.

jjsjann123 commented on tensorflow/contrib/tensorrt/plugin/trt_plugins_test.cc:
```diff
@@ -0,0 +1,112 @@
+#if GOOGLE_CUDA
+#if GOOGLE_TENSORRT
+#include "tensorflow/contrib/tensorrt/plugin/trt_plugin.h"
+#include "tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.h"
```
Just to clarify that NvInfer.h is already inside `#if GOOGLE_TENSORRT` :)

jjsjann123 commented on tensorflow/contrib/tensorrt/plugin/trt_plugins_test.cc:
```diff
@@ -0,0 +1,112 @@
+class StubPlugin : public PluginTensorRT {
+ public:
+  static const std::string plugin_name_;
+  StubPlugin(){};
```
Hmmmm... This one is counterintuitive.
/tensorflow/.clang-format seems to be removing that space. Should I explicitly add it? @aaroey

aaroey replied:
Yes please do.

jjsjann123 commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_utils.cc:
```diff
@@ -14,19 +14,25 @@ limitations under the License.
-  if (serial_length < incremental) return 0;
-  size_t encoded_op_name = *static_cast<const size_t*>(serial_data);
-  return encoded_op_name;
+std::string ExtractOpName(const void* serial_data, size_t serial_length,
```
Assuming TF string and std::string to be API compatible and there is a default conversion, according to:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/types.h#L39

Will update

jjsjann123 committed:
[PR comment addressed]
  Added custom plugin example
    registered tensorflow custom op & plugin kernel
    python wrapper to import custom op & register plugin
  clang-format

jjsjann123 commented:
I didn't enable the custom_plugin_examples as default.
Will do that for following PRs when I wrapped TensorRT official plugins that comes with TRT package (nvplugins)

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -262,6 +271,12 @@ std::vector<int> TFAttrs::get<std::vector<int>>(string key) const {
 }
 
+template <>
+std::vector<float> TFAttrs::get<std::vector<float>>(string key) const {
```
`const string&`

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_kernel.cu.cc:
```diff
@@ -0,0 +1,44 @@
+==============================================================================*/
+
+#include "tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_kernel.h"
+#include "tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_plugin.h"
```
Not needed?

jjsjann123 replied:
I'll rearrange this.

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_kernel.cu.cc:
```diff
@@ -0,0 +1,44 @@
+#include "tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_plugin.h"
+
+#if GOOGLE_CUDA
+#define EIGEN_USE_GPU
```
I think this is needed only when you use eigen. So remove it across this PR?

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/ops/inc_op.cc:
```diff
@@ -0,0 +1,34 @@
+#if GOOGLE_CUDA
+#if GOOGLE_TENSORRT
+
+using namespace tensorflow;
```
Please remove according to https://google.github.io/styleguide/cppguide.html#Namespaces

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/test/plugin_test.py:
```diff
@@ -0,0 +1,93 @@
+from __future__ import division
+from __future__ import print_function
+
+# normally we should do import tensorflow as tf and then
```
Please remove these comments, this is indeed a general convention of TF.

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+from ops import gen_inc_op
+from plugin_wrap import inc_op_register
+from inc_op import *
```
I think we need to disable wildcard-import here.

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/__init__.py:
```diff
@@ -0,0 +1,24 @@
+from plugin_wrap import inc_op_register
+from inc_op import *
+
+# pylint: disable=unused-import,wildcard-import,g-import-not-at-top
```
Remove these comments?

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/__init__.py:
```diff
@@ -0,0 +1,24 @@
+from inc_op import *
+
+# pylint: disable=unused-import,wildcard-import,g-import-not-at-top
+inc_op = gen_inc_op.inc_plugin_trt
```
Does this name conflict with the import? `from inc_op import *`

jjsjann123 replied:
I thought `from inc_op import *` removes the inc_op scope, so we should be good.
But I'll remove the wildcard and change names

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/__init__.py:
```diff
@@ -0,0 +1,24 @@
+# =============================================================================
+"""Import custom op for plugin and register it in plugin factory registry."""
+
+from ops import gen_inc_op
```
Have you tested this code? I suspect we need absolute path for these imports.

jjsjann123 replied:
Yes it is tested and working (at least the part that I need to import ops is working on py27).
Will investigate.

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/BUILD:
```diff
@@ -0,0 +1,110 @@
+    ],
+    deps = if_tensorrt([
+        "@local_config_tensorrt//:nv_infer",
+        "//tensorflow/contrib/tensorrt:trt_plugins",
```
nit: This doesn't need to be inside if_tensorrt, please move it out here and below.

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/test/plugin_test.py:
```diff
@@ -0,0 +1,93 @@
+import numpy as np
+
+# import custom_op as plugin op
+#   the python api handles registration to the plugin factory
```
nit: please fix indentation

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugins_test.cc:
```diff
@@ -0,0 +1,110 @@
+  static const std::string plugin_name_;
+  StubPlugin(){};
+  StubPlugin(const void* serialized_data, size_t length)
+      : PluginTensorRT(serialized_data, length){};
```
Also here.

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugins_test.cc:
```diff
@@ -0,0 +1,110 @@
+  bool RegisterStubPlugin() {
+    if (PluginFactoryTensorRT::GetInstance()->IsPlugin(
+            StubPlugin::plugin_name_))
+      return true;
```
nit: please add parentheses

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin_factory.cc:
```diff
@@ -0,0 +1,79 @@
+                                                    size_t serial_length) {
+  size_t parsed_byte = 0;
+  // extract op_name from serial_data
+  std::string encoded_op_name =
```
Can we use TF string everywhere? Or would you point out to me where do we really need std::string?

jjsjann123 replied:
I should have asked before the changes.
I thought the last request was to use TF string for ExtractOpName only. Will update.

aaroey replied:
Sorry for the confusion. Originally I thought that there may be somewhere in the trt plugin class requires std::string, but I could not find out. Maybe I just missed it, let me know. :)

aaroey commented on tensorflow/contrib/tensorrt/plugin/trt_plugin.cc:
```diff
@@ -0,0 +1,107 @@
+namespace tensorrt {
+
+PluginTensorRT::PluginTensorRT(const void* serialized_data, size_t length) {
+  // sanity check.
```
remove?

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/test/plugin_test.py:
```diff
@@ -0,0 +1,93 @@
+from tensorflow.contrib import tensorrt as trt
+from tensorflow.core.protobuf import config_pb2 as cpb2
+from tensorflow.python.client import session as csess
+from tensorflow.python.framework import dtypes as dtypes
```
I think we don't need `as xxx` if `xxx` is the same as the imported module, so remove?

aaroey replied:
For the others, I think renaming them makes it a bit harder to trace the code, so may be just remove all the renamings?

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/test/plugin_test.py:
```diff
@@ -0,0 +1,93 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Please add a bazel rule to this test. To make it simpler, I think moving the test to the custom_plugin_example dir is sufficient.

aaroey commented on tensorflow/contrib/tensorrt/custom_plugin_examples/inc_op_plugin.h:
```diff
@@ -0,0 +1,107 @@
+#include <cstring>
+#include <iostream>
+#include <string>
+#include <unordered_map>
```
Please remove all unused includes.

aaroey commented:
Hi @jjsjann123,

There seems to be conflicts, would you help to resolve them? After that I'll import this PR to test.
Thanks. :)

jjsjann123 committed:
[PR comment addressed]
  switched from std::string to TF string
  custom_plugin_examples python test added (bazel)
  style guide violation addressed

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into master_pr

resolved conflicts:
	tensorflow/contrib/tensorrt/BUILD

jjsjann123 commented:
Resolved conflicts and comments.
Feel free to ping me if there's issue with internal build. :)

jjsjann123 commented:
Seems like the CI build is complaining about SWIG? Seems to have issue with importing tensorrt and my SWIG interface calling tensorflow::tensorrt::RegisterIncOpPlugin(). Build on local machine is working.
Any hint on how to fix this?

aaroey commented:
Hi @jjsjann123,

Sorry for the delay, I was distracted by some internal trt build problems and is still working on it. Is this PR urgent? If not, may I take a look at it later this week?

Thanks.

aaroey committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into plugin_master_pr

aaroey committed:
Fix build dependency; add missing OpKernel; fix some formatting issues

aaroey committed:
Move/rename the plugin factory test file; delete duplicate test file; fix minor formatting issues.

aaroey committed:
Add a macro for registering the plugin so we don't need to depend on swig; remove the swig file; fix build dependencies; fix tf_custom_op_library by adding GOOGLE_TENSORRT macro when gpu_srcs is not empty.

aaroey committed:
Simplify build dependencies; fix python import order; fix multiple singleton issues by inlining the singleton method.

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into master_pr

jjsjann123 committed:
clang-format

jjsjann123 commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1173,6 +1193,42 @@ tensorflow::Status BinaryTensorOpTensor(
+    // TODO(jie): support only list of float for toy example here.
+    auto data = attrs.get<std::vector<float>>(attr_key);
+    size_t size_data = data.size() * sizeof(float);
+    plugin->SetAttribute(attr_key, static_cast<void*>(data.data()), size_data);
```
This is one of the other placeholders. Technically yes.
Awaiting TRT new plugin API to finalize how we pass attributes & handle serialization/deserialization.

jjsjann123 committed:
detecting SetAttribute failure

aaroey reviewed (changes requested):
Internal build failed, need to fix it.

aaroey committed:
Fix various formatting and build issues.

aaroey committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into plugin_master_pr

aaroey committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into plugin_master_pr

jjsjann123 commented:
Yay~~~
@aaroey Thanks a lot for handling internal build errors and the format issue 💯
